### PR TITLE
Console.Unix: support backspace during Console.Read

### LINKED ISSUE #49129: System.Console.Read() needlessly changes terminal settings on linux
I have noticed that when using Console.Read on linux, the method disables canonical mode (ICANON) flag when reading, may apply to mac too.
Generally while ReadKey reads key by key and to do this it's needed to disable canonical mode and possibly echo when requested, read reads single characters, but actually blocks until newline before starting to return characters, so works similarly to scanf/cin/whatever in other languages.
However, linux itself would handle that just fine in canonical mode, so that the Read call does not have to turn it off. Turning it off makes it impossible to, for example, erase typed characters before pressing return key, or use standard terminal editing features like ctrl+u to erase line.
In canonical mode linux would start returning characters one by one, but only when the user presses enter on a terminal. Pretty sure it's standard unix behavior, and likely same for OSX.
Of course it is also an expected behavior when the program behaves like standard unix tools and just reads from standard input, if user runs it interactively, input is accepted line by line even if the program is not really doing line based input parsing. That way user still has limited edit capability but the program itself still has freedom when it goes to the way it reads input.

### PR DESCRIPTION
**Pull request by tmds**
Fixes https://github.com/dotnet/runtime/issues/49129#issuecomment-796645511.

In addition to the fix, I've refactored the code a bit and added some `Asserts` to help make it more clear.

@stephentoub @eiriktsarpalis ptal

cc @webczat

### CODE REVIEW

File: src/libraries/System.Console/tests/ManualTests/ManualTests.cs [hunk 1]
```diff
@@ -44,6 +44,21 @@ public static void ReadLineFromOpenStandardInput()
             AssertUserExpectedResults("the characters you typed properly echoed as you typed");
         }
 
+        [ConditionalFact(nameof(ManualTestsEnabled))]
+        public static void ConsoleReadSupportsBackspace()
```
Comment (tmds) [inline src/libraries/System.Console/tests/ManualTests/ManualTests.cs:48]: @eiriktsarpalis can you please run this test on Windows and see if it passes?

File: src/libraries/System.Console/src/System/IO/StdInReader.cs [hunk 2]
```diff
@@ -175,37 +177,47 @@ private bool ReadLineCore(bool consumeKeys)
...
+                            {
+                                _readLineSB.Length = len - 1;
+                                removed = true;
+                            }
+                        }
+                        else
                         {
-                            _readLineSB.Length = len - 1;
-                            if (!previouslyProcessed)
+                            removed = _tmpKeys.TryPop(out _);
```
Comment (tmds) [inline src/libraries/System.Console/src/System/IO/StdInReader.cs:192]: This TryPop is the actual bugfix.

File: src/libraries/System.Console/tests/ManualTests/ManualTests.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Console/tests/ManualTests/ManualTests.cs:48]: It fails on my box, with the following message:
```
Please type 'a' 3 times, press 'Backspace' to erase 1, then type a single 'b' and press 'Enter'.
aab
Please press 'a' until it wraps to the next terminal line, then press 'Backspace' until the input is erased, and then type a single 'a' and press 'Enter'.
Input: Use the up, down, left, and right arrow keys to move around.  When done, press enter.
    System.ConsoleManualTests.ConsoleReadSupportsBackspace [FAIL]
      Assert.Equal() Failure
      Expected: 10
      Actual:   13
      Stack Trace:
        C:\Users\eitsarpa\devel\dotnet\runtime\src\libraries\System.Console\tests\ManualTests\ManualTests.cs(56,0): at System.ConsoleManualTests.ConsoleReadSupportsBackspace()
    System.ConsoleManualTests.ReadLine_BackSpaceCanMoveAcrossWrappedLines [FAIL]
      Assert.Equal() Failure
                ↓ (pos 0)
      Expected: a
      Actual:
                ↑ (pos 0)
      Stack Trace:
        C:\Users\eitsarpa\devel\dotnet\runtime\src\libraries\System.Console\tests\ManualTests\ManualTests.cs(70,0): at System.ConsoleManualTests.ReadLine_BackSpaceCanMoveAcrossWrappedLines()
```

Comment (eiriktsarpalis) [reply src/libraries/System.Console/tests/ManualTests/ManualTests.cs:48]: FYI I'm no longer owner of System.Console, cc the current owners @adamsitnik @Jozkee and @carlossanlop.

Comment (tmds) [reply src/libraries/System.Console/tests/ManualTests/ManualTests.cs:48]: > It fails on my box, with the following message:

I pushed a commit that should fix this.

> FYI I'm no longer owner of System.Console, cc the current owners @adamsitnik @Jozkee and @carlossanlop.

ok, thanks for running the test!

Current owners ptal.

**Comment by tmds** [comment]
@adamsitnik @Jozkee and @carlossanlop ptal

**Comment by tmds** [comment]
> @adamsitnik @Jozkee and @carlossanlop ptal

@adamsitnik @Jozkee and @carlossanlop ping

**Comment by jeffhandley** [comment]
I'm sorry for the delay getting this reviewed, @tmds! With the FileStream effort winding down, I expect @adamsitnik, @carlossanlop, and @Jozkee will be able to take a look very soon.

**Comment by jeffhandley** [comment]
Tagging @dotnet/area-system-io. It would be great to get this reviewed before the Preview 6 snap so that we can include this fix in that release.

**Comment by tmds** [comment]
@dotnet/area-system-io ptal

This fixes an issue reported by a user in https://github.com/dotnet/runtime/issues/49129#issuecomment-796645511.

**Comment by jeffhandley** [comment]
@adamsitnik This PR is assigned to you for follow-up/decision before the RC1 snap, with a high priority.

**Review by adamsitnik — APPROVED** [review]
I've tested it on both Windows 10 and Ubuntu 18.04 and all Manual tests are passing. @tmds big thanks for the fix and appologies for such a delay in the review!

**Pull request merged**